tests/unit/test_request_id_header.py

```diff
@@ -0,0 +1,387 @@
+
+    def test_snapshot_read(self):
+        add_select1_result()
+        if not getattr(self.database, "_interceptors", None):
+            self.database._interceptors = MockServerTestBase._interceptors
+        with self.database.snapshot() as snapshot:
+            results = snapshot.execute_sql("select 1")
+            result_list = []
+            for row in results:
+                result_list.append(row)
+                self.assertEqual(1, row[0])
+            self.assertEqual(1, len(result_list))
+
+        requests = self.spanner_service.requests
+        self.assertEqual(2, len(requests), msg=requests)
+        self.assertTrue(isinstance(requests[0], BatchCreateSessionsRequest))
+        self.assertTrue(isinstance(requests[1], ExecuteSqlRequest))
+
+        # Now ensure monotonicity of the received request-id segments.
+        got_stream_segments, got_unary_segments = self.canonicalize_request_id_headers()
+        want_unary_segments = [
+            (
+                "/google.spanner.v1.Spanner/BatchCreateSessions",
+                (1, REQ_RAND_PROCESS_ID, 1, 1, 1, 1),
+            )
+        ]
+        want_stream_segments = [
+            (
+                "/google.spanner.v1.Spanner/ExecuteStreamingSql",
+                (1, REQ_RAND_PROCESS_ID, 1, 1, 2, 1),
+            )
+        ]
+
+        assert got_unary_segments == want_unary_segments
+        assert got_stream_segments == want_stream_segments
+
+    def test_snapshot_read_concurrent(self):
+        def select1():
+            with self.database.snapshot() as snapshot:
+                rows = snapshot.execute_sql("select 1")
+                res_list = []
+                for row in rows:
+                    self.assertEqual(1, row[0])
+                    res_list.append(row)
+                self.assertEqual(1, len(res_list))
+
+        n = 10
+        threads = []
+        for i in range(n):
+            th = threading.Thread(target=select1, name=f"snapshot-select1-{i}")
+            th.run()
+            threads.append(th)
+
+        random.shuffle(threads)
+
+        while True:
+            n_finished = 0
+            for thread in threads:
+                if thread.is_alive():
+                    thread.join()
+                else:
+                    n_finished += 1
+
+            if n_finished == len(threads):
+                break
+
+            time.sleep(1)
+
+        requests = self.spanner_service.requests
+        self.assertEqual(n * 2, len(requests), msg=requests)
+
+        client_id = self.database._nth_client_id
+        channel_id = self.database._channel_id
+        got_stream_segments, got_unary_segments = self.canonicalize_request_id_headers()
+
+        want_unary_segments = [
+            (
+                "/google.spanner.v1.Spanner/BatchCreateSessions",
+                (1, REQ_RAND_PROCESS_ID, client_id, channel_id, 1, 1),
+            ),
+            (
+                "/google.spanner.v1.Spanner/GetSession",
```

Comment:
Note: [This PR](https://github.com/googleapis/python-spanner/pull/1252) should get rid of all of these `GetSession` calls, so rebasing on main my cause this test to fail.